File: 2023/domaci/10-abecedni-poradi/volfmatous/main.cpp
#include <iostream>

bool areWordsInOrder(std::string word1, std::string word2, bool descending) {
    bool letterCaseInOrder = true;
    bool letterCaseInOrderDecided = false;

    int shorterLength = std::min(word1.length(), word2.length());
    for (int i = 0; i < shorterLength; ++i) {
        char char1 = word1[i];
        char char2 = word2[i];

        if (std::tolower(char1) != std::tolower(char2)) {
            bool alphabetAscendingOrder = std::tolower(char1) < std::tolower(char2);
            return descending ? !alphabetAscendingOrder : alphabetAscendingOrder;
        }

        if (letterCaseInOrderDecided) {
            continue;
        }

        if (std::islower(char1) && std::isupper(char2)) {
            letterCaseInOrder = true;
            letterCaseInOrderDecided = true;
            continue;
        }

        if (std::isupper(char1) && std::islower(char2)) {
            letterCaseInOrder = false;
            letterCaseInOrderDecided = true;
            continue;
        }
    }

    if (word1.length() != word2.length()) {
        return word1.length() < word2.length();
    }

    return letterCaseInOrder;
}

int main() {
    int n;
    std::cin >> n;

    bool descending = false;
    bool descendingDecided = false;
    std::string previousWord;
    for (int i = 0; i < n; ++i) {
        std::string currentWord;
        std::cin >> currentWord;

        bool inOrderAscending = areWordsInOrder(previousWord, currentWord, false);
        bool inOrderDescending = areWordsInOrder(previousWord, currentWord, true);

        previousWord = currentWord;

        if (descendingDecided) {
            if ((descending && !inOrderDescending) || (!descending && !inOrderAscending)) {
                std::cout << "unordered" << std::endl;
                return 0;
            }
        } else {
            if (inOrderAscending && inOrderDescending) {
                continue;
            }

            descendingDecided = true;

            if (inOrderAscending) {
                descending = false;
                continue;
            }

            if (inOrderDescending) {
                descending = true;
                continue;
            }

            std::cout << "unordered" << std::endl;
            return 0;
        }
    }

    std::cout << "ordered" << std::endl;
    return 0;
}

File: 2023/domaci/10-abecedni-poradi/volfmatous/main.cs
using System;

internal class Program
{
    static void Main(string[] args)
    {
        int n = int.Parse(Console.ReadLine());

        bool descending = false;
        bool descendingDecided = false;
        string previousWord = string.Empty;
        for (int i = 0; i < n; i++)
        {
            string currentWord = Console.ReadLine();

            bool inOrderAscending = AreWordsInOrder(previousWord, currentWord, false);
            bool inOrderDescending = AreWordsInOrder(previousWord, currentWord, true);

            previousWord = currentWord;

            if (descendingDecided)
            {
                if (descending ? !inOrderDescending : !inOrderAscending)
                {
                    Console.WriteLine("unordered");
                    return;
                }
            }
            else
            {
                if (inOrderAscending && inOrderDescending)
                {
                    continue;
                }

                descendingDecided = true;

                if (inOrderAscending)
                {
                    descending = false;
                    continue;
                }

                if (inOrderDescending)
                {
                    descending = true;
                    continue;
                }

                Console.WriteLine("unordered");
                return;
            }
        }

        Console.WriteLine("ordered");
    }

    static bool AreWordsInOrder(string word1, string word2, bool descending)
    {
        bool letterCaseInOrder = true;
        bool letterCaseInOrderDecided = false;

        int shorterLength = Math.Min(word1.Length, word2.Length);
        for (int i = 0; i < shorterLength; i++)
        {
            char char1 = word1[i];
            char char2 = word2[i];

            if (char.ToLower(char1) != char.ToLower(char2))
            {
                bool alphabetAscendingOrder = char.ToLower(char1) < char.ToLower(char2);
                return descending ? !alphabetAscendingOrder : alphabetAscendingOrder;
            }

            if (letterCaseInOrderDecided)
            {
                continue;
            }

            if (char.IsLower(char1) && char.IsUpper(char2))
            {
                letterCaseInOrder = true;
                letterCaseInOrderDecided = true;
                continue;
            }

            if (char.IsUpper(char1) && char.IsLower(char2))
            {
                letterCaseInOrder = false;
                letterCaseInOrderDecided = true;
                continue;
            }
        }

        if (word1.Length != word2.Length)
        {
            return word1.Length < word2.Length;
        }

        return letterCaseInOrder;
    }
}
